qupath-extension-bioformats/src/main/java/qupath/lib/images/writers/ome/zarr/PyramidalOMEZarrWriter.java

```diff
@@ -0,0 +1,474 @@
+                    }
+            );
+
+            for (int i=1; i<downsamples.size(); i++) {
+                try (ImageServer<BufferedImage> server = new BioFormatsImageServer(
+                        outputPath.toUri(),
+                        "--series",                         // since all level zarr subgroups are already created (see the constructor), BioFormats treat each subgroup
+                        String.valueOf(downsamples.size() - i)  // as a different series, so the series we are interested in must be specified
+                )) {
+                    float progressOffset = (float) i / downsamples.size();
+                    writeLevel(
+                            path,
+                            i,
+                            levels.get(i),
+                            server,
+                            executorService,
+                            progress -> {
+                                if (onProgress != null) {
+                                    onProgress.accept(progressOffset + progress / downsamples.size());
+                                }
+                            }
+                    );
+                }
+
+                root.writeAttributes(new OMEZarrAttributesCreator(new ImageServerMetadata.Builder(server.getMetadata())
+                        .levelsFromDownsamples(downsamples.stream().limit(i+1).mapToDouble(d -> d).toArray())
+                        .build()
+                ).getGroupAttributes());
+            }
+        } finally {
+            executorService.shutdownNow();
+        }
+    }
+
+    /**
+     * Builder to create an instance of a {@link PyramidalOMEZarrWriter}.
+     */
+    public static class Builder {
+
+        private final ImageServer<BufferedImage> server;
+        private final List<Double> downsamples;
+        private Compressor compressor = CompressorFactory.createDefaultCompressor();
+        private int numberOfThreads = ThreadTools.getParallelism();
+        private int maxNumberOfChunks = -1;
+        private int tileWidth = 1024;
+        private int tileHeight = 1024;
+        private ImageRegion boundingBox = null;
+        private int zStart = 0;
+        private int zEnd;
+        private int tStart = 0;
+        private int tEnd;
+
+        /**
+         * Create the builder.
+         *
+         * @param server the image to write. Only the level with downsample 1 will be considered (so it must exist)
+         * @param downsamples the downsamples the output image should have. There must be at least one downsample,
+         *                    and one downsample should be equal to 1
+         * @throws IllegalArgumentException if the downsample list is empty or doesn't contain 1, or if the provided
+         * server doesn't contain a level with downsample 1
+         * @throws NullPointerException if one of the provided parameter is null
+         */
+        public Builder(ImageServer<BufferedImage> server, List<Double> downsamples) {
```

Comment:
Could we use `double... downsamples`?

`List<Double>` is actually hard to get right from Groovy - passing `[1, 4, 16]` gives me errors at runtime because it doesn't automatically convert to `Double`... and `[1.0, 4.0, 16.0]` doesn't work because it uses `BigDecimal`.

Using `double...` is easier and means that `null` isn't possible, but if you really want a list then `List<? extends Number>` could help, then call `doubleValue()` on each element when needed.